### ISSUE TITLE
Dropped messages from kernel after restart

### ISSUE THREAD

**Issue by jasongrout** [jupyter/notebook#3705]
I'm trying to track down a sporadic problem that is cropping up in testing JupyterLab kernel communication.

Essentially, what I'm seeing is on a kernel restart, the kernel/notebook server/browser communication sometimes goes something like this:

Websocket between server and browser is closed immediately, and message buffering starts in the server.
Kernel->server: status busy
Kernel->server: shutdown reply
Kernel->server: status idle

<KERNEL RESTART HAPPENS>
Kernel->server: status starting
<Presumably here is where the notebook server does a kernel info request>
Kernel->server: status busy
Kernel->server: kernel_info_reply
Kernel->server: status idle

<Websocket connection is restored>
Browser->Server: kernel info request
Server->browser: replays the busy/shutdown_reply/idle messages above
Kernel->Server: status busy *but this is not received by the server*
Kernel->Server: kernel info reply
Kernel->Server: status idle *but this is not received by the server*
Server->browser: kernel info reply

So basically the browser is not receiving the status busy/idle messages from its kernel info request.

Any idea what might be happening, or how to debug further?

CC @minrk, @takluyver

**Comment by jasongrout** [comment]
Here is a dump of the logging I've been doing. There's a lot going on here, so I'll try to annotate some:

```
Kernel: restarting (6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47)
closing ws connection: 6e7a63
[D 16:10:25.061 ServicesTestApp] Websocket closed 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47:e8508dcde3edd3b4733ba40fb75ea2d5
[I 16:10:25.061 ServicesTestApp] Starting buffering for 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47:e8508dcde3edd3b4733ba40fb75ea2d5
[D 16:10:25.062 ServicesTestApp] Clearing buffer for 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47
[D 16:10:25.065 ServicesTestApp] Accepting token-authenticated connection from 127.0.0.1
[EchoKernelApp] STATUS - Sent status busy: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 66256, tzinfo=datetime.timezone.utc), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'status', 'msg_id': '8300a0ff-5c0e61b69fc4886842986f70'}, 'msg_id': '8300a0ff-5c0e61b69fc4886842986f70', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'content': {'execution_state': 'busy'}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x106b141d0>}
[EchoKernelApp] 
*** MESSAGE TYPE:shutdown_request***
[EchoKernelApp]    Content: {'restart': True}
   --->
   
[EchoKernelApp] shutdown_request: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f', 'msg_type': 'shutdown_request', 'parent_header': {}, 'metadata': {}, 'content': {'restart': True}, 'buffers': []}
[EchoKernelApp] {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 66941, tzinfo=datetime.timezone.utc), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'shutdown_reply', 'msg_id': '3a0f4b15-531249dfe0f2550bb894d415'}, 'msg_id': '3a0f4b15-531249dfe0f2550bb894d415', 'msg_type': 'shutdown_reply', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'content': {'status': 'ok', 'restart': True}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x106b141d0>}
[EchoKernelApp] STATUS - Sent status idle: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 68480, tzinfo=datetime.timezone.utc), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'status', 'msg_id': 'ab86fadd-f94fe52a7d19988d359fed7a'}, 'msg_id': 'ab86fadd-f94fe52a7d19988d359fed7a', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'content': {'execution_state': 'idle'}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x106b141d0>}
[D 16:10:25.277 ServicesTestApp] Starting kernel: ['/[snip]/miniconda3/envs/devjlab/bin/python', '-m', 'jupyterlab.tests.echo_kernel', '-f', '/var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/jupyter_runtime/kernel-6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47.json', '--debug']
[D 16:10:25.282 ServicesTestApp] Connecting to: tcp://127.0.0.1:53723
[I 16:10:25.283 ServicesTestApp] Kernel restarted: 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47
[D 16:10:25.283 ServicesTestApp] Connecting to: tcp://127.0.0.1:53720
[D 16:10:25.285 ServicesTestApp] Buffering msg on 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47:iopub
[D 16:10:25.285 ServicesTestApp] Buffering msg on 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47:iopub
[D 16:10:25.285 ServicesTestApp] Buffering msg on 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47:iopub
[EchoKernelApp] IPYTHONDIR set to: /var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/ipython
[EchoKernelApp] Using existing profile dir: '/var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/ipython/profile_default'
[EchoKernelApp] Searching path ['/[snip]/projects/app/jupyterlab/packages/services', '/var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/ipython/profile_default', '/[snip]/miniconda3/envs/devjlab/etc/ipython', '/usr/local/etc/ipython', '/etc/ipython'] for config files
[EchoKernelApp] Attempting to load config file: ipython_config.py
[EchoKernelApp] Looking for ipython_config in /etc/ipython
[EchoKernelApp] Looking for ipython_config in /usr/local/etc/ipython
[EchoKernelApp] Looking for ipython_config in /[snip]/miniconda3/envs/devjlab/etc/ipython
[EchoKernelApp] Looking for ipython_config in /var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/ipython/profile_default
[EchoKernelApp] Looking for ipython_config in /[snip]/projects/app/jupyterlab/packages/services
[EchoKernelApp] Attempting to load config file: ipython_kernel_config.py
[EchoKernelApp] Looking for ipython_kernel_config in /etc/ipython
[EchoKernelApp] Looking for ipython_kernel_config in /usr/local/etc/ipython
[EchoKernelApp] Looking for ipython_kernel_config in /[snip]/miniconda3/envs/devjlab/etc/ipython
[EchoKernelApp] Looking for ipython_kernel_config in /var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/ipython/profile_default
[EchoKernelApp] Looking for ipython_kernel_config in /[snip]/projects/app/jupyterlab/packages/services
[EchoKernelApp] Loading connection file /var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/jupyter_runtime/kernel-6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47.json
[EchoKernelApp] Starting the kernel at pid: 31047
[EchoKernelApp] shell ROUTER Channel on port: 53720
[EchoKernelApp] stdin ROUTER Channel on port: 53722
[EchoKernelApp] control ROUTER Channel on port: 53723
[EchoKernelApp] iopub PUB Channel on port: 53721
[EchoKernelApp] Heartbeat REP Channel on port: 53724
[EchoKernelApp] Writing connection file: /var/folders/ln/5623mk393v77n5l3rk2x88h00000gp/T/tmpbkodvoiw/jupyter_runtime/kernel-6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47.json
[EchoKernelApp] To connect another client to this kernel, use:
[EchoKernelApp]     --existing kernel-6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47.json
[EchoKernelApp] STATUS - Sent status starting: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 755766, tzinfo=datetime.timezone.utc), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'status', 'msg_id': 'fc8b7954-822738e9e76a27940bf6790f'}, 'msg_id': 'fc8b7954-822738e9e76a27940bf6790f', 'msg_type': 'status', 'parent_header': {}, 'content': {'execution_state': 'starting'}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x1035962e8>}
[EchoKernelApp] STATUS - Sent status busy: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 805309, tzinfo=datetime.timezone.utc), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'status', 'msg_id': 'ac83894d-3701b7ccb5d64e12ee03be4b'}, 'msg_id': 'ac83894d-3701b7ccb5d64e12ee03be4b', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 284141, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'kernel_info_request', 'msg_id': '01d0c683-ccbe2a950408ff78aea242ae'}, 'content': {'execution_state': 'busy'}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x1035962e8>}
[EchoKernelApp] 
*** MESSAGE TYPE:kernel_info_request***
[EchoKernelApp]    Content: {}
   --->
   
[EchoKernelApp] kernel_info_request: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 284141, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'kernel_info_request', 'msg_id': '01d0c683-ccbe2a950408ff78aea242ae'}, 'msg_id': '01d0c683-ccbe2a950408ff78aea242ae', 'msg_type': 'kernel_info_request', 'parent_header': {}, 'metadata': {}, 'content': {}, 'buffers': []}
[EchoKernelApp] {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 807159, tzinfo=datetime.timezone.utc), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'kernel_info_reply', 'msg_id': 'ecdb624e-55122be5ec62bc15b4a4485e'}, 'msg_id': 'ecdb624e-55122be5ec62bc15b4a4485e', 'msg_type': 'kernel_info_reply', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 284141, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'kernel_info_request', 'msg_id': '01d0c683-ccbe2a950408ff78aea242ae'}, 'content': {'status': 'ok', 'protocol_version': '5.1', 'implementation': 'Echo', 'implementation_version': '1.0', 'language_info': {'name': 'echo', 'mimetype': 'text/plain', 'file_extension': '.txt'}, 'banner': 'Echo kernel - as useful as a parrot', 'help_links': []}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x1035962e8>}
[EchoKernelApp] STATUS - Sent status idle: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 808440, tzinfo=datetime.timezone.utc), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'status', 'msg_id': '2ec54390-f5cf918ca271147fa3c1ea01'}, 'msg_id': '2ec54390-f5cf918ca271147fa3c1ea01', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 284141, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'kernel_info_request', 'msg_id': '01d0c683-ccbe2a950408ff78aea242ae'}, 'content': {'execution_state': 'idle'}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x1035962e8>}
[D 16:10:25.809 ServicesTestApp] Kernel info reply received: 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47
[D 16:10:25.810 ServicesTestApp] 200 POST /api/kernels/6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47/restart?1529622625061 (127.0.0.1) 745.49ms
Kernel: reconnecting (6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47)
Starting WebSocket: ws://localhost:8888/api/kernels/6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47
[D 16:10:25.813 ServicesTestApp] Initializing websocket connection /api/kernels/6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47/channels
[D 16:10:25.814 ServicesTestApp] Accepting token-authenticated connection from 127.0.0.1
[I 16:10:25.815 ServicesTestApp] Adapting to protocol v5.1 for kernel 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47
[D 16:10:25.815 ServicesTestApp] 101 GET /api/kernels/6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47/channels?session_id=e8508dcde3edd3b4733ba40fb75ea2d5&token=3ca46d4126b8b5b5e88215e15cf168701579e9b8c6adbb4c (127.0.0.1) 2.88ms
Kernel: connected (6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47)
[D 16:10:25.815 ServicesTestApp] Opening websocket /api/kernels/6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47/channels
JS KERNEL SENT     MESSAGE: K6e7a63 Mf2004f kernel_info_request
[D 16:10:25.816 ServicesTestApp] Getting buffer for 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47
[I 16:10:25.816 ServicesTestApp] Restoring connection for 6e7a6350-2d1c-44e3-b3d1-88e7ce2d6b47:e8508dcde3edd3b4733ba40fb75ea2d5
[I 16:10:25.816 ServicesTestApp] Replaying 3 buffered messages
[W 16:10:25.817 ServicesTestApp] NBSERVER: KERNEL -> JS: iopub: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 66256, tzinfo=tzutc()), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'status', 'msg_id': '8300a0ff-5c0e61b69fc4886842986f70'}, 'msg_id': '8300a0ff-5c0e61b69fc4886842986f70', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'metadata': {}, 'content': {'execution_state': 'busy'}, 'buffers': []}
[W 16:10:25.817 ServicesTestApp] NBSERVER: SENDING KERNEL -> JS: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 66256, tzinfo=tzutc()), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'status', 'msg_id': '8300a0ff-5c0e61b69fc4886842986f70'}, 'msg_id': '8300a0ff-5c0e61b69fc4886842986f70', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'metadata': {}, 'content': {'execution_state': 'busy'}, 'buffers': []}

IMMEDIATE RECEIVE {"header": {"version": "5.3", "date": "2018-06-21T23:10:25.066256Z", "session": "6bc05d3a-be716bf13c0d9e2494e4ccab", "username": "user", "msg_type": "status", "msg_id": "8300a0ff-5c0e61b69fc4886842986f70"}, "msg_id": "8300a0ff-5c0e61b69fc4886842986f70", "msg_type": "status", "parent_header": {"version": "5.3", "date": "2018-06-21T23:10:25.065395Z", "session": "1b92563a-aeb869eab29059579f8912f9", "username": "user", "msg_type": "shutdown_request", "msg_id": "2b3bf122-baed94cf05e1187909b13c3f"}, "metadata": {}, "content": {"execution_state": "busy"}, "buffers": [], "channel": "iopub"}
JS KERNEL RECEIVED MESSAGE: K6e7a63 Ce8508d KS MS6bc05d P2b3bf1 M8300a0 status busy 
[W 16:10:25.818 ServicesTestApp] NBSERVER: KERNEL -> JS: iopub: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 66941, tzinfo=tzutc()), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'shutdown_reply', 'msg_id': '3a0f4b15-531249dfe0f2550bb894d415'}, 'msg_id': '3a0f4b15-531249dfe0f2550bb894d415', 'msg_type': 'shutdown_reply', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'metadata': {}, 'content': {'status': 'ok', 'restart': True}, 'buffers': []}
[W 16:10:25.818 ServicesTestApp] NBSERVER: SENDING KERNEL -> JS: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 66941, tzinfo=tzutc()), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'shutdown_reply', 'msg_id': '3a0f4b15-531249dfe0f2550bb894d415'}, 'msg_id': '3a0f4b15-531249dfe0f2550bb894d415', 'msg_type': 'shutdown_reply', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'metadata': {}, 'content': {'status': 'ok', 'restart': True}, 'buffers': []}

IMMEDIATE RECEIVE {"header": {"version": "5.3", "date": "2018-06-21T23:10:25.066941Z", "session": "6bc05d3a-be716bf13c0d9e2494e4ccab", "username": "user", "msg_type": "shutdown_reply", "msg_id": "3a0f4b15-531249dfe0f2550bb894d415"}, "msg_id": "3a0f4b15-531249dfe0f2550bb894d415", "msg_type": "shutdown_reply", "parent_header": {"version": "5.3", "date": "2018-06-21T23:10:25.065395Z", "session": "1b92563a-aeb869eab29059579f8912f9", "username": "user", "msg_type": "shutdown_request", "msg_id": "2b3bf122-baed94cf05e1187909b13c3f"}, "metadata": {}, "content": {"status": "ok", "restart": true}, "buffers": [], "channel": "iopub"}
JS KERNEL RECEIVED MESSAGE: K6e7a63 Ce8508d KS6bc05d MS6bc05d P2b3bf1 M3a0f4b shutdown_reply 
[W 16:10:25.818 ServicesTestApp] NBSERVER: KERNEL -> JS: iopub: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 68480, tzinfo=tzutc()), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'status', 'msg_id': 'ab86fadd-f94fe52a7d19988d359fed7a'}, 'msg_id': 'ab86fadd-f94fe52a7d19988d359fed7a', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'metadata': {}, 'content': {'execution_state': 'idle'}, 'buffers': []}
[W 16:10:25.818 ServicesTestApp] NBSERVER: SENDING KERNEL -> JS: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 68480, tzinfo=tzutc()), 'session': '6bc05d3a-be716bf13c0d9e2494e4ccab', 'username': 'user', 'msg_type': 'status', 'msg_id': 'ab86fadd-f94fe52a7d19988d359fed7a'}, 'msg_id': 'ab86fadd-f94fe52a7d19988d359fed7a', 'msg_type': 'status', 'parent_header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 65395, tzinfo=tzutc()), 'session': '1b92563a-aeb869eab29059579f8912f9', 'username': 'user', 'msg_type': 'shutdown_request', 'msg_id': '2b3bf122-baed94cf05e1187909b13c3f'}, 'metadata': {}, 'content': {'execution_state': 'idle'}, 'buffers': []}

IMMEDIATE RECEIVE {"header": {"version": "5.3", "date": "2018-06-21T23:10:25.068480Z", "session": "6bc05d3a-be716bf13c0d9e2494e4ccab", "username": "user", "msg_type": "status", "msg_id": "ab86fadd-f94fe52a7d19988d359fed7a"}, "msg_id": "ab86fadd-f94fe52a7d19988d359fed7a", "msg_type": "status", "parent_header": {"version": "5.3", "date": "2018-06-21T23:10:25.065395Z", "session": "1b92563a-aeb869eab29059579f8912f9", "username": "user", "msg_type": "shutdown_request", "msg_id": "2b3bf122-baed94cf05e1187909b13c3f"}, "metadata": {}, "content": {"execution_state": "idle"}, "buffers": [], "channel": "iopub"}
JS KERNEL RECEIVED MESSAGE: K6e7a63 Ce8508d KS6bc05d MS6bc05d P2b3bf1 Mab86fa status idle 
[W 16:10:25.819 ServicesTestApp] NBSERVER: JS -> KERNEL: {'header': {'username': '', 'version': '5.2', 'session': 'e8508dcde3edd3b4733ba40fb75ea2d5', 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request'}, 'parent_header': {}, 'content': {}, 'metadata': {}, 'buffers': []}
[EchoKernelApp] STATUS - Sent status busy: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 820421, tzinfo=datetime.timezone.utc), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'status', 'msg_id': '105cdc23-cc982e7a030a9de1e21c02e6'}, 'msg_id': '105cdc23-cc982e7a030a9de1e21c02e6', 'msg_type': 'status', 'parent_header': {'username': '', 'version': '5.2', 'session': 'e8508dcde3edd3b4733ba40fb75ea2d5', 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 819643, tzinfo=tzutc())}, 'content': {'execution_state': 'busy'}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x1035962e8>}
[EchoKernelApp] 
*** MESSAGE TYPE:kernel_info_request***
[EchoKernelApp]    Content: {}
   --->
   
[EchoKernelApp] kernel_info_request: {'header': {'username': '', 'version': '5.2', 'session': 'e8508dcde3edd3b4733ba40fb75ea2d5', 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 819643, tzinfo=tzutc())}, 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request', 'parent_header': {}, 'metadata': {}, 'content': {}, 'buffers': []}
[EchoKernelApp] {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 821070, tzinfo=datetime.timezone.utc), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'kernel_info_reply', 'msg_id': '25a7b115-1ec9399bd684549861075650'}, 'msg_id': '25a7b115-1ec9399bd684549861075650', 'msg_type': 'kernel_info_reply', 'parent_header': {'username': '', 'version': '5.2', 'session': 'e8508dcde3edd3b4733ba40fb75ea2d5', 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 819643, tzinfo=tzutc())}, 'content': {'status': 'ok', 'protocol_version': '5.1', 'implementation': 'Echo', 'implementation_version': '1.0', 'language_info': {'name': 'echo', 'mimetype': 'text/plain', 'file_extension': '.txt'}, 'banner': 'Echo kernel - as useful as a parrot', 'help_links': []}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x1035962e8>}
[EchoKernelApp] STATUS - Sent status idle: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 821788, tzinfo=datetime.timezone.utc), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'status', 'msg_id': 'cf48cdc6-59211a7aafe9cf94fc7903fd'}, 'msg_id': 'cf48cdc6-59211a7aafe9cf94fc7903fd', 'msg_type': 'status', 'parent_header': {'username': '', 'version': '5.2', 'session': 'e8508dcde3edd3b4733ba40fb75ea2d5', 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 819643, tzinfo=tzutc())}, 'content': {'execution_state': 'idle'}, 'metadata': {}, 'tracker': <zmq.sugar.tracker.MessageTracker object at 0x1035962e8>}
[W 16:10:25.822 ServicesTestApp] NBSERVER: KERNEL -> JS: shell: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 821070, tzinfo=tzutc()), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'kernel_info_reply', 'msg_id': '25a7b115-1ec9399bd684549861075650'}, 'msg_id': '25a7b115-1ec9399bd684549861075650', 'msg_type': 'kernel_info_reply', 'parent_header': {'username': '', 'version': '5.2', 'session': 'e8508dcde3edd3b4733ba40fb75ea2d5', 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 819643, tzinfo=tzutc())}, 'metadata': {}, 'content': {'status': 'ok', 'protocol_version': '5.1', 'implementation': 'Echo', 'implementation_version': '1.0', 'language_info': {'name': 'echo', 'mimetype': 'text/plain', 'file_extension': '.txt'}, 'banner': 'Echo kernel - as useful as a parrot', 'help_links': []}, 'buffers': []}
[W 16:10:25.822 ServicesTestApp] NBSERVER: SENDING KERNEL -> JS: {'header': {'version': '5.3', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 821070, tzinfo=tzutc()), 'session': 'fb93bba9-8fbdbcb5afeeb6e4e8a57a89', 'username': 'user', 'msg_type': 'kernel_info_reply', 'msg_id': '25a7b115-1ec9399bd684549861075650'}, 'msg_id': '25a7b115-1ec9399bd684549861075650', 'msg_type': 'kernel_info_reply', 'parent_header': {'username': '', 'version': '5.2', 'session': 'e8508dcde3edd3b4733ba40fb75ea2d5', 'msg_id': 'f2004f37a4fdf44466f2b9a5c74cd6a0', 'msg_type': 'kernel_info_request', 'date': datetime.datetime(2018, 6, 21, 23, 10, 25, 819643, tzinfo=tzutc())}, 'metadata': {}, 'content': {'status': 'ok', 'protocol_version': '5.1', 'implementation': 'Echo', 'implementation_version': '1.0', 'language_info': {'name': 'echo', 'mimetype': 'text/plain', 'file_extension': '.txt'}, 'banner': 'Echo kernel - as useful as a parrot', 'help_links': []}, 'buffers': []}

IMMEDIATE RECEIVE {"header": {"version": "5.3", "date": "2018-06-21T23:10:25.821070Z", "session": "fb93bba9-8fbdbcb5afeeb6e4e8a57a89", "username": "user", "msg_type": "kernel_info_reply", "msg_id": "25a7b115-1ec9399bd684549861075650"}, "msg_id": "25a7b115-1ec9399bd684549861075650", "msg_type": "kernel_info_reply", "parent_header": {"username": "", "version": "5.2", "session": "e8508dcde3edd3b4733ba40fb75ea2d5", "msg_id": "f2004f37a4fdf44466f2b9a5c74cd6a0", "msg_type": "kernel_info_request", "date": "2018-06-21T23:10:25.819643Z"}, "metadata": {}, "content": {"status": "ok", "protocol_version": "5.1", "implementation": "Echo", "implementation_version": "1.0", "language_info": {"name": "echo", "mimetype": "text/plain", "file_extension": ".txt"}, "banner": "Echo kernel - as useful as a parrot", "help_links": []}, "buffers": [], "channel": "shell"}
JS KERNEL RECEIVED MESSAGE: K6e7a63 Ce8508d KS6bc05d MSfb93bb Pf2004f M25a7b1 kernel_info_reply 
```
